### PR TITLE
Java object factory: retain tagged types

### PR DESCRIPTION
**Pull request by smowton**
The object factory used to follow all types, meaning if you asked it for a tagged type you'd get a raw struct type back. This alters the factory to keep the tag (desirable when tags can carry generic information), and adds a unit test verifying that it doesn't strip the tag in some common unusual cases (array, enumeration and generic initialisation).

### CODE REVIEW

File: jbmc/src/java_bytecode/java_object_factory.cpp
Comment (thk123) [inline jbmc/src/java_bytecode/java_object_factory.cpp:254]: :pick: Prefer not directly accessing the id:

```C++
if(const auto target_class_type = try_type_dynamic_cast<java_class_type>(followed_target_type))
```

:eagle:

Comment (thk123) [reply jbmc/src/java_bytecode/java_object_factory.cpp:254]: Has the advantage that if we choose to tag java_class_type to differentiate from other struct types, will automatically get checked here.

File: jbmc/src/java_bytecode/java_object_factory.cpp
Comment (thk123) [inline jbmc/src/java_bytecode/java_object_factory.cpp:572]: :eagle:

File: jbmc/src/java_bytecode/java_object_factory.cpp
Comment (thk123) [inline jbmc/src/java_bytecode/java_object_factory.cpp:1057]: :eagle:

**Comment by smowton** [comment]
Test-gen bump (https://github.com/diffblue/test-gen/pull/3036) passing

File: jbmc/unit/java_bytecode/java_object_factory/struct_tag_types.cpp
Comment (thk123) [inline jbmc/unit/java_bytecode/java_object_factory/struct_tag_types.cpp]: :pick: `REQUIRE_FALSE` is a little clearer to read than the `!` IMO

File: jbmc/unit/java_bytecode/java_object_factory/struct_tag_types.cpp
Comment (thk123) [inline jbmc/unit/java_bytecode/java_object_factory/struct_tag_types.cpp:136]: :bulb: You might like to use `require_goto_statements::require_declaration_of_name` either here, or inline, or expand it with your functionality.

File: jbmc/unit/java_bytecode/java_object_factory/struct_tag_types.cpp
Comment (thk123) [inline jbmc/unit/java_bytecode/java_object_factory/struct_tag_types.cpp]: :pick: Here you do a full tree search of the entire ast for code_decl, but in practise it is only `code_blockt` you need to flatten? `require_goto_statements::get_all_statements` does the top level flattening. This has the advantage of not being misleading since a decl can't really appear anywhere in the AST.

File: jbmc/unit/java_bytecode/java_object_factory/struct_tag_types.cpp
Comment (thk123) [inline jbmc/unit/java_bytecode/java_object_factory/struct_tag_types.cpp:36]: :pick: Could these be implemented with a `depth_begin()`/`depth_end()`? Then can use `std::any`?

File: jbmc/unit/java_bytecode/java_object_factory/struct_tag_types.cpp
Comment (thk123) [inline jbmc/unit/java_bytecode/java_object_factory/struct_tag_types.cpp]: :pick: I think this name is misleading: I was expecting it to be the identifier of the declaration, rather than the tag of the type, perhaps a better name would be `contains_decl_with_struct_tag`?

Comment (thk123) [reply jbmc/unit/java_bytecode/java_object_factory/struct_tag_types.cpp]: Also, why do you handle these via just checking tags rather than through the actual type?

**Review by thk123 — APPROVED** [review]
A few nice to haves, but otherwise lgtm

File: jbmc/src/java_bytecode/java_object_factory.cpp
Comment (smowton) [reply jbmc/src/java_bytecode/java_object_factory.cpp:254]: I'll let the `id()` / `dynamic_cast` war work itself out in the other thread before doing this on existing code :)

File: jbmc/unit/java_bytecode/java_object_factory/struct_tag_types.cpp
Comment (smowton) [reply jbmc/unit/java_bytecode/java_object_factory/struct_tag_types.cpp]: Laziness -- they're generics so they have decorators (so they're not `==` to a plain `struct_tag_typet`). Explained in a comment; I'll leave it to the generics tests to verify the structure of generic types.

**Comment by smowton** [comment]
Applied some of the nice-to-haves, will merge once CI passes.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 328bc04aec7df5ffdf6489d9a4345d7709d2087d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100597612

**Pull request merged**
